contracts/PaymentReceiver.sol

```diff
@@ -28,76 +28,114 @@ contract PaymentReceiver is IPaymentReceiver, Database {
     function deleteGate(uint256 _gateId) public override {
         Gate memory gate = getGate(_gateId);
 
-        for (uint256 index = 0; index < gate.activeUsers.length; index++) {
-            address addr = gate.activeUsers[index];
-            _deleteFlow(addr, gate.payee, onlyToken);
+        uint256 index = 0;
+        for(index; index < gate.activeUsers; index++) {
+            address addr = gateUsers[_gateId][index];
+            _deleteFlow(addr, gate.payee, gate.flowRate, onlyToken);
+            checkedIn[addr][_gateId] = false;
             emit CheckOut(addr , _gateId);
-        } 
+        }
+        gate.activeUsers = 0;
 
         super.deleteGate(_gateId);
     }
 
     function checkIn(uint256 _gateId) external {
-        Gate storage gate = gates[_gateId];
+        require(checkedIn[msg.sender][_gateId] == false, "you are already checked in");
 
-        for (uint256 index = 0; index < gate.activeUsers.length; index++) {
-            require(gate.activeUsers[index] != msg.sender, string(abi.encodePacked("already checked in at: ", gate.name)));
-        }
```

Comment:
According to my research reads cost 200 gas, while writes cost either 5000, or 20,000 gas (20,000 if the address written to had no value previously).

The old cost was 
200 - array length read
n * 200 - for every item read
5,000 - updating array length (array.push)
20,000 - writing to a new memory address (array.push), assuming arrays release memory on pop rather than just zeroing it

= 25,200 + n* 200

The new costs are:

200 - initial read in the require
20,000 - `checkedIn[msg.sender][_gateId] = true;` or 5,000 if they are a repeat customer
20,000 `gateUsers[_gateId][gate.activeUsers] = msg.sender` (or, like the array above possibly 5,000)
5000 `gate.activeUsers++`

which gives 45,200

So as long as n <= 100, the original version is cheaper, but basically neither method is too crazy (so happy to merge). 

Does that math check out?